### PR TITLE
Absolutely positioned widgets validation

### PR DESCRIPTION
**Pull request by PolinaShneider**
Added widgets position validation rules:
* Warn if widgets position is out of the grid
* For absolutely positioned widget `width-units` and `height-units` have no effect 

**Config:**
```
[configuration]
  width-units = 10
  height-units = 10

[group]

  [widget]
    type = chart
    position = 11-11
    width-units = 4
    height-units = 3

    [series]
      metric = a
      entity = b

  [widget]
    type = chart

    [series]
      metric = a
      entity = b
```

**Interface:**
Position out of grid warning ⤵️
image

Width-units and height-units (similar) warnings if position is set ⤵️
image

### CODE REVIEW

File: src/relatedSettingsRules/presenceValidation/noUselessSettings/widgetsDimensions.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            .filter(element => element !== undefined);
+
+        const diagnostic: Diagnostic[] = [];
+
+        if (position) {
+            foundSettings.forEach(setting => {
+                diagnostic.push(
+                    createDiagnostic(
+                        setting.textRange,
+                        `\`width-units\` has no effect if \`position\` is specified`,
```
Comment (annstriganova) [inline src/relatedSettingsRules/presenceValidation/noUselessSettings/widgetsDimensions.ts]: Shouldn't it be `${setting.displayName} has no effect if "position" is specified`?

File: src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                     */
+                    outer: for (let i = x1 - 1; i < x2; i++) {
+                        for (let j = y1 - 1; j < y2; j++) {
+                            /**
+                             * We are out of grid defined in configuration
+                             */
+                            if (grid[i] === undefined || grid[i][j] === undefined) {
+                                errors.push(createDiagnostic(
+                                    position.textRange,
+                                    `Widget \`position\` '${position.value}' overflows grid` +
```
Comment (annstriganova) [inline src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: I think it would be better to use double quotes instead of back ticks.

Comment (annstriganova) [reply src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: @VeselovAlex, what do you think about labels?

File: src/relatedSettingsRules/presenceValidation/noUselessSettings/widgetsDimensions.ts (hunk 1 above)
Comment (PolinaShneider) [reply src/relatedSettingsRules/presenceValidation/noUselessSettings/widgetsDimensions.ts]: You are right, corrected this

File: src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts (hunk 2 above)
Comment (PolinaShneider) [reply src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: @unrealwork suggested using code blocks like in section suggestions in HoverProvider.

image

Comment (PolinaShneider) [reply src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: But, frankly speaking, I think it's too much and can be distractive.

Comment (annstriganova) [reply src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: As I remember, markdown is not supported in diagnostic messages.

Comment (PolinaShneider) [reply src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: You are right

File: src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                             * We are out of grid defined in configuration
+                             */
+                            if (grid[i] === undefined || grid[i][j] === undefined) {
+                                errors.push(createDiagnostic(
+                                    position.textRange,
+                                    `Widget ${position.displayName} '${position.value}' overflows grid` +
+                                    ` ${gridHeight} times ${gridWidth}`,
+                                    DiagnosticSeverity.Warning
+                                ));
+                                break outer;
```
Comment (VeselovAlex) [inline src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: I think you should not break this loop here. Fill all occupied cells and then report error on overlapping. It you break here, overlapping with other widgets can be missed

File: src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * Widget position parsing helper function
+ * @param line — position setting
+ */
+function parsePosition(setting: Setting): Coordinates | null {
+    let fullForm: string = setting.value;
+    /**
+     * Process case for 'position = 1-1' shorthand and turn it into 'position = 1-1, 1-1'
+     */
+    if (setting.value.split(",").length === 1) {
```
Comment (VeselovAlex) [inline src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: Isn't it similar to `setting.value.indexOf(',') < 0`?

File: src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    const position = widget.getSettingFromTree("position");
+
+    if (position) {
+        try {
+            const { x1, x2, y1, y2 } = parsePosition(position);
+            if (x1 < 1 || x2 > gridWidth || y1 < 1 || y2 > gridHeight) {
+                return createDiagnostic(
+                    position.textRange,
+                    `Widget ${position.displayName} '${position.value}' overflows grid` +
+                    ` ${gridHeight} times ${gridWidth}`,
```
Comment (VeselovAlex) [inline src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: times -> `×` (&amp;times;)

File: src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts (hunk 4 above)
Comment (VeselovAlex) [reply src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: Also there is no tests for a single position value

File: src/test/widgetsPosition.test.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+[group]
+  [widget]
+    type = chart
+    ${setting}
+    [series]
+      metric = a
+      entity = b`;
+
+suite("Widgets position tests", () => {
+    test("Incorrect widgets position", () => {
```
Comment (VeselovAlex) [inline src/test/widgetsPosition.test.ts]: Some cases:
* `position = 1-1,2-2,3-3`
* `position = 0-0`
* `position = -1-0`
* `position = 1-1 , 2-2 `

File: src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+ * Check if widget position doesn't overflow grid
+ * @param widget - widget section to check
+ */
+function detectGridOverflow(widget: Section): Diagnostic | void {
+    /**
+     * Get grid dimensions out of width- and height-units defined in configuration
+     * Or use default dimensions
+     */
+    const gridWidth = +getValueOfSetting("width-units", widget.parent.parent);
+    const gridHeight = +getValueOfSetting("height-units", widget.parent.parent);
```
Comment (annstriganova) [inline src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: Minor optimization: could it be moved inside `if(position){` branch?

File: src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+ */
+function detectGridOverflow(widget: Section): Diagnostic | void {
+    const position = widget.getSettingFromTree("position");
+
+    if (position) {
+        /**
+         * Get grid dimensions out of width- and height-units defined in configuration
+         * Or use default dimensions
+         */
+        const gridWidth = +getValueOfSetting("width-units", widget.parent.parent);
```
Comment (VeselovAlex) [inline src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: Shouldn't we perform NaN check?

File: src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        const gridHeight = +getValueOfSetting("height-units", widget.parent.parent);
+
+        try {
+            const { x1, x2, y1, y2 } = parsePosition(position);
+            if (x1 < 1 || x2 > gridWidth || y1 < 1 || y2 > gridHeight) {
+                return createDiagnostic(
+                    position.textRange,
+                    `Widget ${position.displayName} '${position.value}' overflows grid` +
+                    ` ${gridHeight} × ${gridWidth}`,
```
Comment (VeselovAlex) [inline src/relatedSettingsRules/valueValidation/widgetsGridOverflow.ts]: IMHO whitespace around &times; is unnecessary

**Pull request merged**